iocuddle-sgx/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![deny(clippy::all)]
+#![allow(clippy::identity_op)]
+
+use std::marker::PhantomData;
+
+use bitflags::bitflags;
+use iocuddle::*;
+use sgx_types::{page, secs, sig};
+
+const SGX: Group = Group::new(0xA4);
+pub const ENCLAVE_CREATE: Ioctl<Write, &Create> = unsafe { SGX.write(0x00) };
+pub const ENCLAVE_ADD_PAGES: Ioctl<WriteRead, &AddPages> = unsafe { SGX.write_read(0x01) };
+pub const ENCLAVE_INIT: Ioctl<Write, &Init> = unsafe { SGX.write(0x02) };
+pub const ENCLAVE_SET_ATTRIBUTE: Ioctl<Write, &SetAttribute> = unsafe { SGX.write(0x03) };
+
+bitflags! {
+    pub struct Flags: u64 {
+        const MEASURE = 1 << 0;
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct Create<'a>(u64, PhantomData<&'a ()>);
```

Comment:
Do you mind summing up the rationale behind this lifetime maybe giving an example?

IIUC this binds the lifetime of `Create` to the `&secs` passed in to `new`?

Reply:
Correct. Casting the incoming pointer to a `u64` erases lifetime information. We add that information manually using the `PhantomData`. This ensures that `Create` lives at least as long as `&secs`. We need this to be correct since we're passing this pointer to the kernel as a `u64`.

This is going to be a common idiom in pretty much all ioctls. Do you want me to document it every time? Just once? Or should we just explain it if questions arise?

Reply:
FYI, I don't think this usage is really any different than the one summarized in the official documentation for `PhantomData`: https://doc.rust-lang.org/std/marker/struct.PhantomData.html#unused-lifetime-parameters

The only difference is that we're using `u64` instead of a pointer.